// php-wasm/PhpBase.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PhpBase = void 0;
var _config = require("./config.js");
var _OutputBuffer = require("./OutputBuffer.js");
var _Event2 = require("./_Event.js");
var _fsOps = require("./fsOps.js");
var _resolveDependencies2 = require("./resolveDependencies.js");
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t["return"] || t["return"](); } finally { if (u) throw o; } } }; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _wrapNativeSuper(t) { var r = "function" == typeof Map ? new Map() : void 0; return _wrapNativeSuper = function _wrapNativeSuper(t) { if (null === t || !_isNativeFunction(t)) return t; if ("function" != typeof t) throw new TypeError("Super expression must either be null or a function"); if (void 0 !== r) { if (r.has(t)) return r.get(t); r.set(t, Wrapper); } function Wrapper() { return _construct(t, arguments, _getPrototypeOf(this).constructor); } return Wrapper.prototype = Object.create(t.prototype, { constructor: { value: Wrapper, enumerable: !1, writable: !0, configurable: !0 } }), _setPrototypeOf(Wrapper, t); }, _wrapNativeSuper(t); }
function _construct(t, e, r) { if (_isNativeReflectConstruct()) return Reflect.construct.apply(null, arguments); var o = [null]; o.push.apply(o, e); var p = new (t.bind.apply(t, o))(); return r && _setPrototypeOf(p, r.prototype), p; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _isNativeFunction(t) { try { return -1 !== Function.toString.call(t).indexOf("[native code]"); } catch (n) { return "function" == typeof t; } }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
var STR = 'string';
var NUM = 'number';
var PhpBase = exports.PhpBase = /*#__PURE__*/function (_EventTarget) {
  function PhpBase(PhpBinary) {
    var _globalThis$phpSettin;
    var _this;
    var args = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var sapi = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'embed';
    _classCallCheck(this, PhpBase);
    _this = _callSuper(this, PhpBase);
    _this.queue = [];
    _this.onerror = function () {};
    _this.onoutput = function () {};
    _this.onready = function () {};
    Object.defineProperty(_this, 'encoder', {
      value: new TextEncoder()
    });
    Object.defineProperty(_this, 'buffers', {
      value: {
        stdin: [],
        stdout: new _OutputBuffer.OutputBuffer(_this, 'output', -1),
        stderr: new _OutputBuffer.OutputBuffer(_this, 'error', -1)
      }
    });
    Object.freeze(_this.buffers);
    _this.autoTransaction = 'autoTransaction' in args ? args.autoTransaction : true;
    _this.transactionStarted = false;
    _this.shared = args.shared = 'shared' in args ? args.shared : {};
    _this.phpArgs = args;
    var defaults = {
      stdin: function stdin() {
        var _this$buffers$stdin$s;
        return (_this$buffers$stdin$s = _this.buffers.stdin.shift()) !== null && _this$buffers$stdin$s !== void 0 ? _this$buffers$stdin$s : null;
      },
      stdout: function stdout(_byte) {
        return _this.buffers.stdout.push(_byte);
      },
      stderr: function stderr(_byte2) {
        return _this.buffers.stderr.push(_byte2);
      },
      postRun: function postRun() {
        var event = new _Event2._Event('ready');
        _this.onready(event);
        _this.dispatchEvent(event);
      }
    };
    var fixed = {
      onRefresh: new Set()
    };
    var phpSettings = (_globalThis$phpSettin = globalThis.phpSettings) !== null && _globalThis$phpSettin !== void 0 ? _globalThis$phpSettin : {};
    var userLocateFile = args.locateFile || function () {
      return undefined;
    };
    var files = args.files || [];
    var _resolveDependencies = (0, _resolveDependencies2.resolveDependencies)(args.sharedLibs, _this),
      extraFiles = _resolveDependencies.files,
      libs = _resolveDependencies.libs,
      urlLibs = _resolveDependencies.urlLibs;
    args.locateFile = function (path, directory) {
      var located = userLocateFile(path, directory);
      if (located !== undefined) {
        return located;
      }
      if (urlLibs[path]) {
        return urlLibs[path];
      }
    };
    _this.valueIndex = 0;
    _this.binary = new PhpBinary(Object.assign({}, defaults, phpSettings, args, fixed)).then(/*#__PURE__*/function () {
      var _ref = _asyncToGenerator(function* (php) {
        php.ccall('pib_storage_init', NUM, [], []);
        if (!php.FS.analyzePath('/preload').exists) {
          php.FS.mkdir('/preload');
        }
        yield Promise.all(files.concat(extraFiles).map(function (fileDef) {
          return new Promise(function (accept) {
            return php.FS.createPreloadedFile(fileDef.parent, fileDef.name, fileDef.url, true, false, accept);
          });
        }));
        var iniLines = libs.map(function (lib) {
          if (typeof lib === 'string' || lib instanceof URL) {
            return "extension=".concat(lib);
          } else if (_typeof(lib) === 'object' && lib.ini) {
            return "extension=".concat(String(lib.url).split('/').pop());
          }
        });
        args.ini && iniLines.push(args.ini.replace(/\n\s+/g, '\n'));
        php.FS.writeFile('/php.ini', iniLines.join("\n") + "\n", {
          encoding: 'utf8'
        });
        yield php.ccall('pib_init', NUM, [STR], [sapi], {
          async: true
        });
        return php;
      });
      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
    return _this;
  }
  _inherits(PhpBase, _EventTarget);
  return _createClass(PhpBase, [{
    key: "inputString",
    value: function inputString(byteString) {
      this.input(this.encoder.encode(byteString));
    }
  }, {
    key: "input",
    value: function input(items) {
      var _this$buffers$stdin;
      (_this$buffers$stdin = this.buffers.stdin).push.apply(_this$buffers$stdin, _toConsumableArray(items));
    }
  }, {
    key: "flush",
    value: function flush() {
      this.buffers.stdout.flush();
      this.buffers.stderr.flush();
    }
  }, {
    key: "tokenize",
    value: function tokenize(phpCode) {
      return this.binary.then(function (php) {
        return php.ccall('pib_tokenize', STR, [STR], [phpCode]);
      });
    }
  }, {
    key: "startTransaction",
    value: function startTransaction() {
      return Promise.resolve();
    }
  }, {
    key: "commitTransaction",
    value: function commitTransaction() {
      var readOnly = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
      return Promise.resolve();
    }
  }, {
    key: "_enqueue",
    value: function () {
      var _enqueue2 = _asyncToGenerator(function* (callback) {
        var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
        var readOnly = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
        var accept, reject;
        var coordinator = new Promise(function (a, r) {
          var _ref2;
          return _ref2 = [a, r], accept = _ref2[0], reject = _ref2[1], _ref2;
        });
        var _accept = function _accept(result) {
          return accept(result);
        };
        var _reject = function _reject(reason) {
          return reject(reason);
        };
        this.queue.push([callback, params, _accept, _reject]);
        if (!this.queue.length) {
          return;
        }
        (yield this.autoTransaction && !readOnly) ? this.startTransaction() : Promise.resolve();
        while (this.queue.length) {
          var _this$queue$shift = this.queue.shift(),
            _this$queue$shift2 = _slicedToArray(_this$queue$shift, 4),
            _callback = _this$queue$shift2[0],
            _params = _this$queue$shift2[1],
            _accept2 = _this$queue$shift2[2],
            _reject2 = _this$queue$shift2[3];
          yield _callback.apply(void 0, _toConsumableArray(_params)).then(_accept2)["catch"](_reject2);
        }
        (yield this.autoTransaction) ? this.commitTransaction(readOnly) : Promise.resolve();
        return coordinator;
      });
      function _enqueue(_x2) {
        return _enqueue2.apply(this, arguments);
      }
      return _enqueue;
    }()
  }, {
    key: "run",
    value: function run(phpCode) {
      var _this2 = this;
      return this._enqueue(function (phpCode) {
        return _this2._run(phpCode);
      }, [phpCode]);
    }
  }, {
    key: "_run",
    value: function () {
      var _run2 = _asyncToGenerator(function* (phpCode) {
        var _this3 = this;
        var call = (yield this.binary).ccall('pib_run', NUM, [STR], ["?>".concat(phpCode)], {
          async: true
        });
        return call["finally"](function () {
          return _this3.flush();
        });
      });
      function _run(_x3) {
        return _run2.apply(this, arguments);
      }
      return _run;
    }()
  }, {
    key: "exec",
    value: function exec(phpCode) {
      var _this4 = this;
      return this._enqueue(function (phpCode) {
        return _this4._exec(phpCode);
      }, [phpCode]);
    }
  }, {
    key: "_exec",
    value: function () {
      var _exec2 = _asyncToGenerator(function* (phpCode) {
        var _this5 = this;
        var call = (yield this.binary).ccall('pib_exec', STR, [STR], [phpCode], {
          async: true
        });
        return call["finally"](function () {
          return _this5.flush();
        });
      });
      function _exec(_x4) {
        return _exec2.apply(this, arguments);
      }
      return _exec;
    }()
  }, {
    key: "x",
    value: function () {
      var _x6 = _asyncToGenerator(function* (fragments) {
        var names = [];
        var phpModule = yield this.binary;
        for (var _len = arguments.length, values = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          values[_key - 1] = arguments[_key];
        }
        if (phpModule.hasVrzno) {
          var _iterator = _createForOfIteratorHelper(values),
            _step;
          try {
            for (_iterator.s(); !(_step = _iterator.n()).done;) {
              var value = _step.value;
              var name = "___value__".concat(this.valueIndex++);
              this.shared[name] = value;
              names.push(name);
            }
          } catch (err) {
            _iterator.e(err);
          } finally {
            _iterator.f();
          }
          var code = '';
          fragments = _toConsumableArray(fragments);
          while (fragments.length || names.length) {
            if (fragments.length) code += fragments.shift();
            if (names.length) {
              code += "(vrzno_shared('".concat(names.shift(), "'))");
            }
          }
          code = "vrzno_zval( ".concat(code, " );");
          return phpModule.zvalToJS(yield this.exec(code));
        } else {
          var encoded = values.map(function (value) {
            return JSON.stringify(value);
          });
          fragments = _toConsumableArray(fragments);
          var _code = '';
          while (fragments.length || names.length) {
            if (fragments.length) _code += fragments.shift();
            if (encoded.length) {
              _code += "(json_decode('".concat(encoded.shift(), "'))");
            }
          }
          return this.exec(_code);
        }
      });
      function x(_x5) {
        return _x6.apply(this, arguments);
      }
      return x;
    }()
  }, {
    key: "r",
    value: function () {
      var _r = _asyncToGenerator(function* (fragments) {
        var names = [];
        var phpModule = yield this.binary;
        for (var _len2 = arguments.length, values = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
          values[_key2 - 1] = arguments[_key2];
        }
        if (phpModule.hasVrzno) {
          var _iterator2 = _createForOfIteratorHelper(values),
            _step2;
          try {
            for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
              var value = _step2.value;
              var name = "___value__".concat(this.valueIndex++);
              this.shared[name] = value;
              names.push(name);
            }
          } catch (err) {
            _iterator2.e(err);
          } finally {
            _iterator2.f();
          }
          var code = '';
          fragments = _toConsumableArray(fragments);
          while (fragments.length || names.length) {
            if (fragments.length) code += fragments.shift();
            if (names.length) {
              code += "(vrzno_shared('".concat(names.shift(), "'))");
            }
          }
          return this.run(code);
        } else {
          var encoded = values.map(function (value) {
            return JSON.stringify(value);
          });
          fragments = _toConsumableArray(fragments);
          var _code2 = '';
          while (fragments.length || names.length) {
            if (fragments.length) _code2 += fragments.shift();
            if (encoded.length) {
              _code2 += "(json_decode('".concat(encoded.shift(), "'))");
            }
          }
          return this.run(_code2);
        }
      });
      function r(_x7) {
        return _r.apply(this, arguments);
      }
      return r;
    }()
  }, {
    key: "refresh",
    value: function () {
      var _refresh = _asyncToGenerator(function* () {
        var _this6 = this;
        var php = yield this.binary;
        var _iterator3 = _createForOfIteratorHelper(php.onRefresh),
          _step3;
        try {
          for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
            var callback = _step3.value;
            callback();
          }
        } catch (err) {
          _iterator3.e(err);
        } finally {
          _iterator3.f();
        }
        Object.keys(this.shared).forEach(function (key) {
          return delete _this6.shared[key];
        });
        return php.ccall('pib_refresh', NUM, [], [], {
          async: true
        });
      });
      function refresh() {
        return _refresh.apply(this, arguments);
      }
      return refresh;
    }()
  }, {
    key: "analyzePath",
    value: function analyzePath(path) {
      return this._enqueue(_fsOps.fsOps.analyzePath, [this.binary, path]);
    }
  }, {
    key: "readdir",
    value: function readdir(path) {
      return this._enqueue(_fsOps.fsOps.readdir, [this.binary, path]);
    }
  }, {
    key: "readFile",
    value: function readFile(path, options) {
      return this._enqueue(_fsOps.fsOps.readFile, [this.binary, path, options]);
    }
  }, {
    key: "stat",
    value: function stat(path) {
      return this._enqueue(_fsOps.fsOps.stat, [this.binary, path]);
    }
  }, {
    key: "mkdir",
    value: function mkdir(path) {
      return this._enqueue(_fsOps.fsOps.mkdir, [this.binary, path]);
    }
  }, {
    key: "rmdir",
    value: function rmdir(path) {
      return this._enqueue(_fsOps.fsOps.rmdir, [this.binary, path]);
    }
  }, {
    key: "rename",
    value: function rename(path, newPath) {
      return this._enqueue(_fsOps.fsOps.rename, [this.binary, path, newPath]);
    }
  }, {
    key: "writeFile",
    value: function writeFile(path, data, options) {
      return this._enqueue(_fsOps.fsOps.writeFile, [this.binary, path, data, options]);
    }
  }, {
    key: "unlink",
    value: function unlink(path) {
      return this._enqueue(_fsOps.fsOps.unlink, [this.binary, path]);
    }
  }]);
}(/*#__PURE__*/_wrapNativeSuper(EventTarget));
PhpBase.phpVersion = _config.phpVersion;
PhpBase.phpVersionFull = _config.phpVersionFull;
